rfd/headless-login.md

```diff
@@ -0,0 +1,365 @@
+    // PubKey is a public key user wishes to sign in
+    PubKey []byte `json:"pub_key"`
+    // Command is the client command being requested
+    Command string
+    // State is the request state (pending, approved, or denied)
+    State string
+}
+```
+
+These request details are saved on the Auth server in a `HeadlessLoginRequest` object, with the backend prefix `/headless_login_requests/`. They will be saved with a 1 minute TTL, by which point the user should have completed the authentication flow.
+
+In response, the client receives a request ID generated by the server, which it will use to generate a client command or URL to begin local authentication: `tsh headless approve <request_id>` and `https://proxy.example.com/headless/<request_id>/approve`.
+
+The command and URL are shared with the user so they can complete local authentication for the request in a local terminal or web browser.
+
+#### Local authentication
+
+When the user runs the command or opens the URL locally, their local login session will be used to connect to the Teleport Auth server. If the user is not yet connected, they will be prompted to login with MFA as usual.
+
+Once connected, the user can view the request details and either approve or deny the request:
+
+* If the user approves the request, they will need to pass an MFA challenge to update the request to the approved state.
+* If the user denies the request, the request will be updated to the denied state.
+
+#### Certificate retrieval
+
+Once the headless login request is approved, the remote client can request MFA-verified Single-user (1 minute TTL) user certificates from the Auth server. In this request, the user will be challenged to sign for the public key saved in the request details on the Auth Server to confirm the client's identity.
+
+The resulting user certificates will then be used by the client to complete the `tsh` request initially requested, e.g. `tsh ssh user@node01`.
+
+### Audit log
+
+The following actions will be tracked with audit events:
+
+* User initiates headless login
+* User approves/denies headless login request
+* User receives headless login credentials
+
+### Server changes
+
+Headless authentication has a unique API flow compared to other login methods.
+
+#### `POST /webapi/login/headless/begin`
+
+This endpoint is used to initiate headless login and get a request ID. Like other login endpoints, this endpoint is not authenticated and can be called by anyone with access to the Teleport Proxy address.
+
+```go
+type BeginHeadlessLoginRequest struct {
+    // User is a teleport username
+    User string `json:"user"`
+    // PubKey is a public key user wishes to sign
+    PubKey []byte
+    // Command is the client command being requested
+    Command string
+}
+
+type BeginHeadlessLoginResponse struct {
+    // RequestID is a uuid associated with the headless login request
+    RequestID string `json:"user"`
+}
+```
+
+#### `rpc GetHeadlessLoginRequest`
```

Comment:
👍 for the endpoints. I think I would be good to mention them in `Headless authentication overview` diagram.